types.go

```diff
@@ -0,0 +1,7 @@
+package middleware
+
+import "net/http"
+
+type Provider interface {
+	Middleware(http.Handler) http.Handler
+}
```

Comment:
I don't see a purpose for this interface. It is not consumed anywhere in this package (which is a [common anti-pattern](https://github.com/golang/go/wiki/CodeReviewComments#interfaces)), and is redundant with the [internal interface](https://github.com/gorilla/mux/blob/98cb6bf42e086f6af920b965c38cacc07402d51b/middleware.go#L13-L16) consumed by gorilla/mux.

These middleware only need to implement `Middleware(handler http.Handler) http.Handler` as a gorilla/mux convenience, and should return concrete pointer or struct types.